core/src/apps/cardano/sign_tx.py

```diff
@@ -171,68 +147,39 @@ def _process_outputs(self):
     def _build_witnesses(self, tx_aux_hash: str):
         witnesses = []
         for input in self.inputs:
-            _, node = derive_address_and_node(self.keychain, input.address_n)
-            message = (
-                b"\x01" + cbor.encode(self.protocol_magic) + b"\x58\x20" + tx_aux_hash
-            )
-            signature = ed25519.sign_ext(
-                node.private_key(), node.private_key_ext(), message
-            )
-            extended_public_key = (
-                remove_ed25519_prefix(node.public_key()) + node.chain_code()
-            )
-            witnesses.append(
-                [
-                    (input.type or 0),
-                    cbor.Tagged(24, cbor.encode([extended_public_key, signature])),
-                ]
-            )
-
-        return witnesses
-
-    @staticmethod
-    def compute_fee(input_coins_sum: int, outgoing_coins: list, change_coins: list):
-        outgoing_coins_sum = sum(outgoing_coins)
-        change_coins_sum = sum(change_coins)
-
-        return input_coins_sum - outgoing_coins_sum - change_coins_sum
+            node = self.keychain.derive(input.address_n)
+            message = b"\x58\x20" + tx_aux_hash
+
+            signature = ed25519.sign_ext(node.private_key(), node.private_key_ext(), message)
+
+            public_key = remove_ed25519_prefix(node.public_key())
+
+            witnesses.append([public_key, signature])
+
+        return {0: witnesses}
 
     def serialise_tx(self):
 
         self._process_outputs()
 
         inputs_cbor = []
         for input in self.inputs:
-            inputs_cbor.append(
-                [
-                    (input.type or 0),
-                    cbor.Tagged(24, cbor.encode([input.prev_hash, input.prev_index])),
-                ]
-            )
-
-        inputs_cbor = cbor.IndefiniteLengthArray(inputs_cbor)
+            inputs_cbor.append([input.prev_hash, input.prev_index])
 
         outputs_cbor = []
         for index, address in enumerate(self.output_addresses):
-            outputs_cbor.append(
-                [cbor.Raw(base58.decode(address)), self.outgoing_coins[index]]
-            )
+            outputs_cbor.append([unhexlify(address), self.outgoing_coins[index]])
 
         for index, address in enumerate(self.change_addresses):
-            outputs_cbor.append(
-                [cbor.Raw(base58.decode(address)), self.change_coins[index]]
-            )
+            outputs_cbor.append([address, self.change_coins[index]])
 
-        outputs_cbor = cbor.IndefiniteLengthArray(outputs_cbor)
+        # todo: certificates, withdrawals, metadata
+        tx_body = {0: inputs_cbor, 1: outputs_cbor, 2: self.fee, 3: self.ttl}
 
-        tx_aux_cbor = [inputs_cbor, outputs_cbor, self.attributes]
-        tx_hash = hashlib.blake2b(data=cbor.encode(tx_aux_cbor), outlen=32).digest()
+        tx_body_cbor = cbor.encode(tx_body)
+        tx_hash = hashlib.blake2b(data=tx_body_cbor, outlen=32).digest()
```

Comment:
Does it make sense to make this 32 a symbolic constant? Or even better, move the hash function into a separate method (generally, this makes code more testable --- one can replace an opaque hash function by a transparent one (i.e. one that is easy to invert) to track what goes into hashes --- not that we do need it here).

Reply:
I'll think about this later. Maybe it'll be a part of the file rewrite.